File: flutter-boilerplate/lib/lib/rtc/rtc_telemetry.dart
import '../activity_logger.dart';

/// Sends RTC diagnostics through the same ActivityLogger pipeline used by the
/// app-wide Flutter exception and page telemetry.
void logRtcEvent({
  required String event,
  required String rtcKind,
  String? rtcId,
  String? roomName,
  String? mediaType,
  String? phase,
  String? exceptionType,
  Object? error,
  StackTrace? stackTrace,
  Map<String, dynamic>? metadata,
}) {
  ActivityLogger.instance.enqueue({
    'category': 'rtc',
    'event': event,
    'eventType': event,
    if (rtcKind.isNotEmpty) 'rtcKind': rtcKind,
    if (rtcId != null && rtcId.isNotEmpty) 'rtcId': rtcId,
    if (roomName != null && roomName.isNotEmpty) 'roomName': roomName,
    if (mediaType != null) 'mediaType': mediaType,
    if (phase != null) 'phase': phase,
    if (exceptionType != null) 'exceptionType': exceptionType,
    if (error != null) 'errorMessage': error.toString(),
    if (stackTrace != null) 'stack': stackTrace.toString(),
    if (metadata != null) 'metadata': metadata,
  });
}

File: tanstack-boilerplate/src/lib/rtc/rtc-telemetry.ts
import { eventLogger } from "@/lib/event-logger";
import type { RtcTelemetryEventOptions } from "@/types/rtc/RtcTelemetry-types";

function errorFields(error: unknown): {
  errorMessage?: string;
  stack?: string;
} {
  if (error instanceof Error) {
    return {
      errorMessage: error.message,
      stack: error.stack,
    };
  }
  if (error === undefined || error === null) return {};
  return { errorMessage: String(error) };
}

/** Emits RTC telemetry through the same batched client-event pipeline as the rest of the app. */
export function logRtcEvent(options: RtcTelemetryEventOptions): void {
  const {
    event,
    error,
    metadata,
    rtcKind,
    rtcId,
    roomName,
    mediaType,
    phase,
    exceptionType,
  } = options;
  const fields = errorFields(error);

  eventLogger.emit({
    eventType: event,
    event,
    category: "rtc",
    url: window.location.pathname,
    userAgent: navigator.userAgent,
    rtcKind,
    ...(rtcId ? { rtcId } : {}),
    ...(roomName ? { roomName } : {}),
    ...(mediaType ? { mediaType } : {}),
    ...(phase ? { phase } : {}),
    ...fields,
    ...(exceptionType ? { exceptionType } : {}),
    ...(metadata ? { metadata } : {}),
  });
}
